les/clientpool.go

```diff
@@ -103,296 +106,180 @@ type clientPool struct {
+	allowInactive() bool
 }
 
-// clientInfo represents a connected client
+// clientInfo defines all information required by clientpool.
 type clientInfo struct {
-	address                string
-	id                     enode.ID
-	connectedAt            mclock.AbsTime
-	capacity               uint64
-	priority               bool
-	pool                   *clientPool
-	peer                   clientPoolPeer
-	queueIndex             int // position in connectedQueue
-	balanceTracker         balanceTracker
-	posFactors, negFactors priceFactors
-	balanceMetaInfo        string
+	node        *enode.Node
+	address     string
+	peer        clientPoolPeer
+	connected   bool
+	connectedAt mclock.AbsTime
+	balance     *lps.NodeBalance
 }
 
-// connSetIndex callback updates clientInfo item index in connectedQueue
-func connSetIndex(a interface{}, index int) {
-	a.(*clientInfo).queueIndex = index
-}
-
-// connPriority callback returns actual priority of clientInfo item in connectedQueue
-func connPriority(a interface{}, now mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	return c.balanceTracker.getPriority(now)
-}
-
-// connMaxPriority callback returns estimated maximum priority of clientInfo item in connectedQueue
-func connMaxPriority(a interface{}, until mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	pri := c.balanceTracker.estimatedPriority(until, true)
-	c.balanceTracker.addCallback(balanceCallbackQueue, pri+1, func() {
-		c.pool.lock.Lock()
-		if c.queueIndex != -1 {
-			c.pool.connectedQueue.Update(c.queueIndex)
+// newClientPool creates a new client pool
+func newClientPool(lespayDb ethdb.Database, minCap, freeClientCap uint64, activeBias time.Duration, clock mclock.Clock, removePeer func(enode.ID)) *clientPool {
+	if minCap > freeClientCap {
+		panic(nil)
+	}
+	ns := nodestate.NewNodeStateMachine(nil, nil, clock, clientPoolSetup)
+	pool := &clientPool{
+		ns:                  ns,
+		BalanceTrackerSetup: balanceTrackerSetup,
+		PriorityPoolSetup:   priorityPoolSetup,
+		clock:               clock,
+		minCap:              minCap,
+		freeClientCap:       freeClientCap,
+		removePeer:          removePeer,
+	}
+	pool.ti = lps.NewTokenIssuer(ns, tokenIssuerSetup, clock, freeClientCap, activeBias/8, time.Hour)
+	pool.posExp = pool.ti.NewTokenExpirer()
+	pool.negExp = pool.ti.NewTokenExpirer()
+	pool.bt = lps.NewBalanceTracker(ns, balanceTrackerSetup, lespayDb, clock, pool.posExp, pool.negExp)
+	pool.ti.SetTotalAmountCallback(pool.bt.TotalTokenAmount)
+	pool.pp = lps.NewPriorityPool(ns, priorityPoolSetup, clock, minCap, activeBias, 4)
+
+	// set default expiration constants used by tests
+	// Note: server overwrites this if token sale is active
+	pool.bt.SetExpirationTCs(0, defaultNegExpTC)
+	// calculate total token balance amount
+
+	ns.SubscribeState(pool.InactiveFlag.Or(pool.PriorityFlag), func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if newState.Equals(pool.InactiveFlag) {
+			ns.AddTimeout(node, pool.InactiveFlag, inactiveTimeout)
```

Comment:
Why adding the timeout here?

Reply:
The concept is that inactive connections are only allowed for a limited time for free clients (so that we can't end up with a ton of idle TCP connections). It should be just enough to do price negotiation and buy tokens (it has to happen through TCP if UDP is not working for some reason). For paid clients it is better to allow staying connected because if a paid client is deactivated that is only a temporary situation and the client will be reactivated soon.